File: backend/python/utils/math_helpers.py
"""
Mathematical Helper Functions for Financial Risk Simulations - Python Implementation

This module provides core mathematical utilities used across different
simulation strategies including random number generation, statistical
functions, and financial calculations.
"""

import numpy as np
import scipy.stats as stats
from typing import List, Tuple, Dict, Any, Optional
import warnings

def random_normal(mean: float = 0, std: float = 1, size: Optional[int] = None) -> np.ndarray:
    """
    Generate random number(s) from normal distribution
    
    Args:
        mean: Mean of the distribution (default: 0)
        std: Standard deviation (default: 1)
        size: Number of samples (None for single value)
    
    Returns:
        Random number(s) from normal distribution
    """
    return np.random.normal(mean, std, size)

def calculate_percentile(arr: np.ndarray, percentile: float) -> float:
    """
    Calculate percentile of an array
    
    Args:
        arr: Array of numbers
        percentile: Percentile to calculate (0-100)
    
    Returns:
        Value at the specified percentile
    """
    return np.percentile(arr, percentile)

def calculate_var(losses: np.ndarray, confidence: float = 0.95) -> float:
    """
    Calculate Value at Risk (VaR) from a distribution of losses
    
    Args:
        losses: Array of loss values
        confidence: Confidence level (e.g., 0.95 for 95%)
    
    Returns:
        VaR value
    """
    percentile_value = confidence * 100
    return calculate_percentile(losses, percentile_value)

def calculate_cvar(losses: np.ndarray, confidence: float = 0.95) -> float:
    """
    Calculate Conditional Value at Risk (CVaR) - Expected Shortfall
    
    Args:
        losses: Array of loss values
        confidence: Confidence level (e.g., 0.95 for 95%)
    
    Returns:
        CVaR value
    """
    var_threshold = calculate_var(losses, confidence)
    tail_losses = losses[losses >= var_threshold]
    
    if len(tail_losses) == 0:
        return var_threshold
    
    return np.mean(tail_losses)

def calculate_portfolio_stats(weights: np.ndarray, 
                            expected_returns: np.ndarray, 
                            covariance_matrix: np.ndarray) -> Dict[str, float]:
    """
    Calculate portfolio statistics (mean, variance, volatility)
    
    Args:
        weights: Portfolio weights (must sum to 1)
        expected_returns: Expected returns for each asset
        covariance_matrix: Covariance matrix
    
    Returns:
        Dictionary with portfolio statistics
    """
    # Portfolio expected return: w^T * μ
    portfolio_return = np.dot(weights, expected_returns)
    
    # Portfolio variance: w^T * Σ * w
    portfolio_variance = np.dot(weights.T, np.dot(covariance_matrix, weights))
    portfolio_volatility = np.sqrt(portfolio_variance)
    
    return {
        'expected_return': float(portfolio_return),
        'variance': float(portfolio_variance),
        'volatility': float(portfolio_volatility)
    }

def correlation_to_covariance(volatilities: np.ndarray, 
                            correlation_matrix: np.ndarray) -> np.ndarray:
    """
    Create covariance matrix from volatilities and correlation coefficients
    
    Args:
        volatilities: Array of asset volatilities
        correlation_matrix: Correlation matrix
    
    Returns:
        Covariance matrix
    """
    vol_matrix = np.diag(volatilities)
    return np.dot(vol_matrix, np.dot(correlation_matrix, vol_matrix))

def cholesky_decomposition(matrix: np.ndarray) -> np.ndarray:
    """
    Perform Cholesky decomposition for correlated random variables
    
    Args:
        matrix: Positive definite matrix
    
    Returns:
        Lower triangular matrix L where A = L * L^T
    """
    try:
        return np.linalg.cholesky(matrix)
    except np.linalg.LinAlgError:
        # If matrix is not positive definite, try to fix it
        eigenvals, eigenvecs = np.linalg.eigh(matrix)
        eigenvals = np.maximum(eigenvals, 1e-8)  # Ensure positive eigenvalues
        fixed_matrix = np.dot(eigenvecs, np.dot(np.diag(eigenvals), eigenvecs.T))
        return np.linalg.cholesky(fixed_matrix)

def generate_correlated_normals(size: int, covariance_matrix: np.ndarray) -> np.ndarray:
    """
    Generate correlated random variables using Cholesky decomposition
    
    Args:
        size: Number of samples to generate
        covariance_matrix: Covariance matrix
    
    Returns:
        Array of correlated random vectors
    """
    n = covariance_matrix.shape[0]
    
    # Generate independent standard normals
    independent = np.random.standard_normal((size, n))
    
    # Cholesky decomposition
    L = cholesky_decomposition(covariance_matrix)
    
    # Transform to correlated variables: Y = X * L^T
    correlated = np.dot(independent, L.T)
    
    return correlated

def calculate_statistics(data: np.ndarray) -> Dict[str, float]:
    """
    Calculate basic statistical measures
    
    Args:
        data: Array of numerical data
    
    Returns:
        Dictionary with statistical measures
    """
    if len(data) == 0:
        return {
            'mean': 0.0, 'std': 0.0, 'min': 0.0, 'max': 0.0, 
            'median': 0.0, 'variance': 0.0, 'count': 0
        }
    
    return {
        'mean': float(np.mean(data)),
        'std': float(np.std(data)),
        'variance': float(np.var(data)),
        'min': float(np.min(data)),
        'max': float(np.max(data)),
        'median': float(np.median(data)),
        'count': len(data)
    }

def bootstrap_sample(data: np.ndarray, sample_size: Optional[int] = None) -> np.ndarray:
    """
    Bootstrap sampling - sample with replacement
    
    Args:
        data: Original data array
        sample_size: Size of bootstrap sample (default: same as original)
    
    Returns:
        Bootstrap sample
    """
    if sample_size is None:
        sample_size = len(data)
    
    indices = np.random.choice(len(data), size=sample_size, replace=True)
    return data[indices]

def normal_cdf(z: float) -> float:
    """
    Standard normal cumulative distribution function
    
    Args:
        z: Z-score
    
    Returns:
        Cumulative probability
    """
    return stats.norm.cdf(z)

def normal_inverse_cdf(p: float) -> float:
    """
    Inverse normal cumulative distribution function
    
    Args:
        p: Probability (0 < p < 1)
    
    Returns:
        Z-score
    """
    return stats.norm.ppf(p)

def calculate_skewness(data: np.ndarray) -> float:
    """
    Calculate skewness of a data series
    
    Args:
        data: Data array
    
    Returns:
        Skewness value
    """
    return float(stats.skew(data))

def calculate_kurtosis(data: np.ndarray) -> float:
    """
    Calculate kurtosis of a data series
    
    Args:
        data: Data array
    
    Returns:
        Kurtosis value
    """
    return float(stats.kurtosis(data))

def calculate_autocorrelation(data: np.ndarray, max_lag: int = 5) -> List[float]:
    """
    Calculate autocorrelation for different lags
    
    Args:
        data: Time series data
        max_lag: Maximum lag to calculate
    
    Returns:
        List of autocorrelation values
    """
    n = len(data)
    data_centered = data - np.mean(data)
    autocorrelations = []
    
    for lag in range(1, max_lag + 1):
        if lag >= n:
            autocorrelations.append(0.0)
            continue
            
        numerator = np.sum(data_centered[:-lag] * data_centered[lag:])
        denominator = np.sum(data_centered ** 2)
        
        correlation = numerator / denominator if denominator > 0 else 0.0
        autocorrelations.append(float(correlation))
    
    return autocorrelations

def calculate_correlation(series1: np.ndarray, series2: np.ndarray) -> float:
    """
    Calculate correlation between two return series
    
    Args:
        series1: First return series
        series2: Second return series
    
    Returns:
        Correlation coefficient
    """
    return float(np.corrcoef(series1, series2)[0, 1])

def validate_portfolio(portfolio: List[float]) -> np.ndarray:
    """
    Validate and convert portfolio to numpy array
    
    Args:
        portfolio: List of portfolio values
    
    Returns:
        Validated numpy array
    
    Raises:
        ValueError: If portfolio is invalid
    """
    portfolio_array = np.array(portfolio)
    
    if len(portfolio_array) == 0:
        raise ValueError("Portfolio cannot be empty")
    
    if np.any(portfolio_array <= 0):
        raise ValueError("All portfolio values must be positive")
    
    if not np.all(np.isfinite(portfolio_array)):
        raise ValueError("Portfolio values must be finite numbers")
    
    return portfolio_array

def validate_correlation_matrix(corr_matrix: np.ndarray) -> np.ndarray:
    """
    Validate correlation matrix
    
    Args:
        corr_matrix: Correlation matrix to validate
    
    Returns:
        Validated correlation matrix
    
    Raises:
        ValueError: If matrix is invalid
    """
    if corr_matrix.shape[0] != corr_matrix.shape[1]:
        raise ValueError("Correlation matrix must be square")
    
    if not np.allclose(corr_matrix, corr_matrix.T):
        raise ValueError("Correlation matrix must be symmetric")
    
    if not np.allclose(np.diag(corr_matrix), 1.0):
        raise ValueError("Correlation matrix diagonal must be 1.0")
    
    if np.any(np.abs(corr_matrix) > 1.0):
        raise ValueError("Correlation values must be between -1 and 1")
    
    # Check if positive semi-definite
    eigenvals = np.linalg.eigvals(corr_matrix)
    if np.any(eigenvals < -1e-8):
        warnings.warn("Correlation matrix is not positive semi-definite")
    
    return corr_matrix

def ensure_positive_definite(matrix: np.ndarray, min_eigenval: float = 1e-8) -> np.ndarray:
    """
    Ensure matrix is positive definite by adjusting eigenvalues
    
    Args:
        matrix: Input matrix
        min_eigenval: Minimum eigenvalue to enforce
    
    Returns:
        Positive definite matrix
    """
    eigenvals, eigenvecs = np.linalg.eigh(matrix)
    eigenvals = np.maximum(eigenvals, min_eigenval)
    return np.dot(eigenvecs, np.dot(np.diag(eigenvals), eigenvecs.T))

File: backend/node/utils/mathHelpers.js
/**
 * Mathematical Helper Functions for Financial Risk Simulations
 * 
 * This module provides core mathematical utilities used across different
 * simulation strategies including random number generation, statistical
 * functions, and financial calculations.
 */

const math = require('mathjs');

/**
 * Generate random number from normal distribution using Box-Muller transform
 * 
 * @param {number} mean - Mean of the distribution (default: 0)
 * @param {number} std - Standard deviation (default: 1)
 * @returns {number} Random number from normal distribution
 */
function randomNormal(mean = 0, std = 1) {
    // Box-Muller transformation for generating normal random variables
    let u1 = 0, u2 = 0;
    // Ensure we don't get log(0)
    while (u1 === 0) u1 = Math.random();
    while (u2 === 0) u2 = Math.random();

    const randStdNormal = Math.sqrt(-2.0 * Math.log(u1)) * Math.cos(2.0 * Math.PI * u2);
    return mean + std * randStdNormal;
}

/**
 * Generate array of random normal numbers
 * 
 * @param {number} size - Number of random numbers to generate
 * @param {number} mean - Mean of the distribution
 * @param {number} std - Standard deviation
 * @returns {Array<number>} Array of random normal numbers
 */
function randomNormalArray(size, mean = 0, std = 1) {
    const result = [];
    for (let i = 0; i < size; i++) {
        result.push(randomNormal(mean, std));
    }
    return result;
}

/**
 * Calculate percentile of an array
 * 
 * @param {Array<number>} arr - Sorted array of numbers
 * @param {number} percentile - Percentile to calculate (0-100)
 * @returns {number} Value at the specified percentile
 */
function calculatePercentile(arr, percentile) {
    if (arr.length === 0) return 0;

    const sorted = [...arr].sort((a, b) => a - b);
    const index = (percentile / 100) * (sorted.length - 1);

    if (Math.floor(index) === index) {
        return sorted[index];
    } else {
        const lower = sorted[Math.floor(index)];
        const upper = sorted[Math.ceil(index)];
        return lower + (upper - lower) * (index - Math.floor(index));
    }
}

/**
 * Calculate Value at Risk (VaR) from a distribution of losses
 * 
 * @param {Array<number>} losses - Array of loss values
 * @param {number} confidence - Confidence level (e.g., 0.95 for 95%)
 * @returns {number} VaR value
 */
function calculateVaR(losses, confidence = 0.95) {
    const percentileValue = confidence * 100;
    return calculatePercentile(losses, percentileValue);
}

/**
 * Calculate Conditional Value at Risk (CVaR) - Expected Shortfall
 * 
 * @param {Array<number>} losses - Array of loss values
 * @param {number} confidence - Confidence level (e.g., 0.95 for 95%)
 * @returns {number} CVaR value
 */
function calculateCVaR(losses, confidence = 0.95) {
    const sorted = [...losses].sort((a, b) => a - b);
    const varIndex = Math.floor((1 - confidence) * sorted.length);

    // CVaR is the average of losses beyond VaR
    if (varIndex === 0) return sorted[0];

    const tailLosses = sorted.slice(0, varIndex);
    return tailLosses.reduce((sum, loss) => sum + loss, 0) / tailLosses.length;
}

/**
 * Calculate portfolio statistics (mean, variance, volatility)
 * 
 * @param {Array<number>} weights - Portfolio weights (must sum to 1)
 * @param {Array<number>} expectedReturns - Expected returns for each asset
 * @param {Array<Array<number>>} covarianceMatrix - Covariance matrix
 * @returns {Object} Portfolio statistics
 */
function calculatePortfolioStats(weights, expectedReturns, covarianceMatrix) {
    // Portfolio expected return: w^T * μ
    const portfolioReturn = weights.reduce((sum, weight, i) =>
        sum + weight * expectedReturns[i], 0);

    // Portfolio variance: w^T * Σ * w
    let portfolioVariance = 0;
    for (let i = 0; i < weights.length; i++) {
        for (let j = 0; j < weights.length; j++) {
            portfolioVariance += weights[i] * weights[j] * covarianceMatrix[i][j];
        }
    }

    const portfolioVolatility = Math.sqrt(portfolioVariance);

    return {
        expectedReturn: portfolioReturn,
        variance: portfolioVariance,
        volatility: portfolioVolatility
    };
}

/**
 * Create correlation matrix from volatilities and correlation coefficients
 * 
 * @param {Array<number>} volatilities - Array of asset volatilities
 * @param {Array<Array<number>>} correlationMatrix - Correlation matrix
 * @returns {Array<Array<number>>} Covariance matrix
 */
function correlationToCovariance(volatilities, correlationMatrix) {
    const n = volatilities.length;
    const covarianceMatrix = [];

    for (let i = 0; i < n; i++) {
        covarianceMatrix[i] = [];
        for (let j = 0; j < n; j++) {
            covarianceMatrix[i][j] = volatilities[i] * volatilities[j] * correlationMatrix[i][j];
        }
    }

    return covarianceMatrix;
}

/**
 * Perform Cholesky decomposition for correlated random variables
 * 
 * @param {Array<Array<number>>} matrix - Positive definite matrix
 * @returns {Array<Array<number>>} Lower triangular matrix L where A = L * L^T
 */
function choleskyDecomposition(matrix) {
    const n = matrix.length;
    const L = Array(n).fill().map(() => Array(n).fill(0));

    for (let i = 0; i < n; i++) {
        for (let j = 0; j <= i; j++) {
            if (i === j) {
                // Diagonal elements
                let sum = 0;
                for (let k = 0; k < j; k++) {
                    sum += L[i][k] * L[i][k];
                }
                L[i][j] = Math.sqrt(matrix[i][i] - sum);
            } else {
                // Non-diagonal elements
                let sum = 0;
                for (let k = 0; k < j; k++) {
                    sum += L[i][k] * L[j][k];
                }
                L[i][j] = (matrix[i][j] - sum) / L[j][j];
            }
        }
    }

    return L;
}

/**
 * Generate correlated random variables using Cholesky decomposition
 * 
 * @param {number} size - Number of samples to generate
 * @param {Array<Array<number>>} covarianceMatrix - Covariance matrix
 * @returns {Array<Array<number>>} Array of correlated random vectors
 */
function generateCorrelatedNormals(size, covarianceMatrix) {
    const n = covarianceMatrix.length;
    const L = choleskyDecomposition(covarianceMatrix);
    const result = [];

    for (let i = 0; i < size; i++) {
        // Generate independent standard normals
        const independent = randomNormalArray(n, 0, 1);

        // Transform to correlated variables: Y = L * X
        const correlated = Array(n).fill(0);
        for (let j = 0; j < n; j++) {
            for (let k = 0; k <= j; k++) {
                correlated[j] += L[j][k] * independent[k];
            }
        }

        result.push(correlated);
    }

    return result;
}

/**
 * Calculate basic statistical measures
 * 
 * @param {Array<number>} data - Array of numerical data
 * @returns {Object} Statistical measures (mean, std, min, max, etc.)
 */
function calculateStatistics(data) {
    if (data.length === 0) {
        return { mean: 0, std: 0, min: 0, max: 0, median: 0 };
    }

    const sorted = [...data].sort((a, b) => a - b);
    const mean = data.reduce((sum, val) => sum + val, 0) / data.length;

    const variance = data.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / data.length;
    const std = Math.sqrt(variance);

    const median = sorted.length % 2 === 0
        ? (sorted[sorted.length / 2 - 1] + sorted[sorted.length / 2]) / 2
        : sorted[Math.floor(sorted.length / 2)];

    return {
        mean,
        std,
        variance,
        min: sorted[0],
        max: sorted[sorted.length - 1],
        median,
        count: data.length
    };
}

/**
 * Bootstrap sampling - sample with replacement
 * 
 * @param {Array<number>} data - Original data array
 * @param {number} sampleSize - Size of bootstrap sample
 * @returns {Array<number>} Bootstrap sample
 */
function bootstrapSample(data, sampleSize = data.length) {
    const sample = [];
    for (let i = 0; i < sampleSize; i++) {
        const randomIndex = Math.floor(Math.random() * data.length);
        sample.push(data[randomIndex]);
    }
    return sample;
}

module.exports = {
    randomNormal,
    randomNormalArray,
    calculatePercentile,
    calculateVaR,
    calculateCVaR,
    calculatePortfolioStats,
    correlationToCovariance,
    choleskyDecomposition,
    generateCorrelatedNormals,
    calculateStatistics,
    bootstrapSample
};
